src/coreclr/src/zap/zapinfo.cpp

```diff
@@ -2147,10 +2147,17 @@ DWORD FilterNamedIntrinsicMethodAttribs(ZapInfo* pZapInfo, DWORD attribs, CORINF
         // is because they often change the code they emit based on what ISAs are supported by the compiler,
         // but we don't know what the target machine will support.
         //
-        // Additionally, we make sure none of the hardware intrinsic method bodies get pregenerated in crossgen
+        // Additionally, we make sure none of the hardware intrinsic method bodies (except ARM64) get pregenerated in crossgen
         // (see ZapInfo::CompileMethod) but get JITted instead. The JITted method will have the correct
         // answer for the CPU the code is running on.
-        fTreatAsRegularMethodCall = (fIsGetIsSupportedMethod && fIsPlatformHWIntrinsic) || (!fIsPlatformHWIntrinsic && fIsHWIntrinsic);
+
+        // For Arm64, AdvSimd/ArmBase is the baseline that is suported and hence we do pregenerate the method bodies
+        // of ARM4 harware intrinsic.
+        fTreatAsRegularMethodCall = (fIsGetIsSupportedMethod && fIsPlatformHWIntrinsic);
+#if !defined(TARGET_ARM64)
+        fTreatAsRegularMethodCall |= (!fIsPlatformHWIntrinsic && fIsHWIntrinsic);
```

Comment:
I thought we determined it would actually be fine for these to just work on x86 as well (but maybe @CarolEidt, @BruceForstall, and @davidwrighton could confirm)?

Essentially, in the case where the tiered JIT is disabled, having the calls like `Create`, `GetLower`, and `ToScalar` just use the SSE2 equivalents is better (and matches what you would get on SSE2 hardware anyways).

When the tiered JIT is available, then rejit should end up improving the codegen after ~30 iterations anyways and it will improve startup time.
The downside, as I understand, is then that rejit doesn't necessarily replace the method body immediately and so you might not get the newer SSE41 or AVX compiled code until later.

Reply:
I'll leave that for @CarolEidt and @davidwrighton .

I'll note a nit (typo) while I'm here: "of ARM4 harware intrinsic" => "of ARM64 hardware intrinsics"

Reply:
It's my recollection as well that we said these should just work on x86 as well, though I recall that we had some discussion of the fact that x86 is notably different in that there's a greater variation of available features.
@davidwrighton - what is your recollection?

Reply:
I just remembered the logic here. This wouldn't be safe. We compile System.Private.CoreLib with the instruction set such that it thinks its compiling with support for SSE4.1 and such. However, we disabled the non-platform intrinsics as usages of those are not wrapped in IsSupported checks. Do not do this.

Reply:
Is this true even though we check for the baseline ISA here: https://github.com/dotnet/runtime/blob/master/src/coreclr/src/jit/hwintrinsicxarch.cpp#L718-L735 (and in managed for the recursive case: https://source.dot.net/#System.Private.CoreLib/Vector128.cs,265)
and then we do opportunistic checks later in here: https://github.com/dotnet/runtime/blob/master/src/coreclr/src/jit/lowerxarch.cpp#L1497, https://github.com/dotnet/runtime/blob/master/src/coreclr/src/jit/lowerxarch.cpp#L1605, etc?

So wouldn't the method be flagged as requiring SSE2 and opportunistically supporting SSE3, SSE41, or AVX in both cases? (I'm probably missing something, the rules are a bit hard to keep straight 😄).

Reply:
Yes. The opportunistic check's are only opportunistic on Crossgen2. On Crossgen1 they are normal checks. This general problem is one that will fix itself when we move to crossgen2 as the default R2R compiler, and that won't happen till .NET 6.